src/main/java/com/firebolt/jdbc/connection/settings/FireboltProperties.java

```diff
@@ -84,35 +88,63 @@ public static FireboltProperties of(Properties... properties) {
 		int tcpKeepIdle = getSetting(mergedProperties, FireboltSessionProperty.TCP_KEEP_IDLE);
 		int tcpKeepCount = getSetting(mergedProperties, FireboltSessionProperty.TCP_KEEP_COUNT);
 		boolean logResultSet = getSetting(mergedProperties, FireboltSessionProperty.LOG_RESULT_SET);
+		String configuredEnvironment = getSetting(mergedProperties, FireboltSessionProperty.ENVIRONMENT);
 		String driverVersions = getSetting(mergedProperties, FireboltSessionProperty.USER_DRIVERS);
 		String clientVersions = getSetting(mergedProperties, FireboltSessionProperty.USER_CLIENTS);
 
-		String host = getHost(mergedProperties);
+		String environment = getEnvironment(configuredEnvironment, mergedProperties);
+		String host = getHost(configuredEnvironment, mergedProperties);
 		Integer port = getPort(mergedProperties, ssl);
-		String database = getDatabase(mergedProperties, path);
+
 		Map<String, String> additionalProperties = getFireboltCustomProperties(mergedProperties);
 
 		return FireboltProperties.builder().ssl(ssl).sslCertificatePath(sslRootCertificate).sslMode(sslMode).path(path)
-				.port(port).database(database).compress(compress).user(user).password(password).host(host)
+				.port(port).database(database).compress(compress).principal(principal).secret(secret).host(host)
 				.additionalProperties(additionalProperties).account(account).engine(engine)
 				.keepAliveTimeoutMillis(keepAliveMillis).maxConnectionsTotal(maxTotal).maxRetries(maxRetries)
 				.bufferSize(bufferSize).socketTimeoutMillis(socketTimeout).connectionTimeoutMillis(connectionTimeout)
 				.tcpKeepInterval(tcpKeepInterval).tcpKeepCount(tcpKeepCount).tcpKeepIdle(tcpKeepIdle)
 				.logResultSet(logResultSet).systemEngine(isSystemEngine)
+				.environment(environment)
 				.userDrivers(driverVersions)
 				.userClients(clientVersions)
 				.build();
 	}
 
-	private static String getHost(Properties properties) {
+	private static String getEngine(Properties mergedProperties, String database) {
+		String engine = getSetting(mergedProperties, FireboltSessionProperty.ENGINE);
+		if (StringUtils.isEmpty(engine) && StringUtils.isEmpty(database)) {
+			return SYSTEM_ENGINE_NAME;
```

Comment:
Generally null is bad. It can produce NullPointerException. Surely, it is possible but it is a little bit dangerous, so I'd suggest to postpone this. I'd change API to return Optional<String> instead but it is too much for now IMHO.

Reply:
As mentioned above, we would need to remove any mention of the term `system engine`, so I would suggest to simply return an empty string here.

Reply:
Now this function is extremely simple. It returns engine as is. For system engine it returns `null`.